### ISSUE TITLE
request, add secondary signal inlets to clip~

### ISSUE THREAD

**Issue by porres** [pure-data/pure-data#2107]
I don't see why it should not include it, all the other max like objects take signals in the secondary inlets.

**Comment by sebshader** [comment]
if so, imo it should be like the audio math objects where arguments control whether they are control or signal-rate.
But of course it's trivial to accomplish the same with ```[min~]``` and ```[max~]```, and the goal isn't really to be max..

**Comment by porres** [comment]
> it should be like the audio math objects where arguments control whether they are control or signal-rate.

that'd be consistent

**Comment by porres** [comment]
even though lop~ got a 2nd sig inlet aware without this

**Comment by Spacechild1** [comment]
> imo it should be like the audio math objects where arguments control whether they are control or signal-rate.

Since Pd 0.54 this is not necessarily true anymore! Objects can now be registered with `CLASS_NOPROMOTESIG`, so that scalars are not automatically promoted to signals; instead you must check in the "dsp" method whether you have a signal or scalar and use to the appropriate perform routine accordingly.

What I find confusing is that many existing objects, such as `[+~]`, continue to force scalar rate if an argument is provided, but some objects, most notably `[osc~]`, `[phasor~]` and `[lop~]`, do not (anymore).

**Comment by porres** [comment]
> What I find confusing is that many existing objects, such as [+~], continue to force scalar rate if an argument is provided, but some objects, most notably [osc~], [phasor~] and [lop~], do not (anymore).

At one point long ago it was the case that [osc~] and [phasor~] would only work with float input if initialized with an argument. I only know that because I cleaned up the documentation and it used to say so, meaning that the docs weren't updated. But then, the case here is that it just promotes floats to signals, unlike [lop~] (and this is why I mentioned it, I know there was a recent change for it and it seems this is the only one that uses it)

I see it makes no sense to give argument when you want a signal in, but it does work for [osc~], [phasor~], some other objects we may also look for (raw filters like [rpole~] come to mind) and now [lop~].

So yeah, I guess we can make it like [lop~] for [clip~], and I could open a new ticket to make it so that binops~ be like [lop~]

**Comment by Spacechild1** [comment]
> But then, the case here is that it just promotes floats to signals, unlike [lop~]

Ah, you're right! `[osc~]` and `[phasor~]` do not have a scalar perform routine at all.

**Comment by porres** [comment]
I'm working on it and it looks good, but it behaves bad if only one of the secondary inlets is connected, I guess I have to make 4 perform routines, no signal, signal a, signal b and both signals

**Comment by porres** [comment]
> I guess I have to make 4 perform routines, no signal, signal a, signal b and both signals

got here just fine, but I hit a wall now trying to figure out how to make it wrap multichannel signals like binops~ maybe I can make a draft PR and @Spacechild1 can help me elaborate? :)

It's working fine though for multichannel signal inputs of the same size

**Comment by porres** [comment]
Ok, still thinking about how to deal with channels of different sizes. 

I'm not sure it makes a whole lot of sense to apply the exact same procedure of binary operators, in which case it takes the signal with most channels and then sets the object to output the same number of channels and then it takes the other smaller signal and repeats and wraps it til it reaches the number of the other bigger signal.

clip~ has 3 inlets and the left input is what we expect to come out with the same number of channels, right?

As another example, I see [fft~] does not use the same logic of binary operators, so let's just keep that logic for them only. I also see that [fft~] warns if the number of channels mismatch! Of course it makes a lot of sense in this case. For [clip~] I am thinking of something that makes a lot of sense for it.

If you have a multichannel connection in the left inlet and secondary inlets are just a single channel, then they are copied and used to clip the left input. Now, if the secondary inlets have more than a single channel, then it needs to be the same size of the mutichannel left input, otherwise we have an error. This is also quite simple to implement, what do you think?

**Comment by umlaeute** [comment]
as discussed in #2009: why not just use the wrap-channels method when the numbers of (right-hand signals) channels do not match?
for the "valid" cases (either there's only 1 channel or exactly the same number as the left-hand signal) this would give you the desired output.
for the other cases (where there's no "natural" mapping), it would give you a result. (probably an unexpected, but that might be interesting anyhow; more importantly we get a consistent wrapping pattern across various objects).

you might warn (though probably only once), instead of erroring out, if you prefer that.

**Comment by porres** [comment]
> why not just use the wrap-channels method when the numbers of (right-hand signals) channels do not match?

I made my points earlier, I think this kinda works well for binops like I said, but this looks really crazy to me with 3 inlets and weird to take care of, say, 3 channel left input and then 2 and 5 channels for the secondary inlets... 

Also, I have to admit, I was having enough of a hard time trying to figure out how to use the same code and wrap just for 2 inputs, let alone 3...

I won't really oppose to the idea if you all think it's best and makes sense, but it seems I won't be able to  implement this myself, if @umlaeute or @Spacechild1 would like you could ignore my 2nd commit where I do this to implement how to deal with channels mismatch and help me with the desired way.

@millerpuckette could of course also join in :)

cheers
